### PR TITLE
Properly resolve MRO entires for subclassed methods

### LINKED ISSUE #19: autodoc: Inherited async methods not detected
When using autodoc to document a class that inherits an async method from its base class, the inherited method is not detected as *async*.

Basically, with code like this:

```
class SomeBaseClass:
    """The base class."""
    async def base_async_method(self):
        """This is an async method in the base class."""
        pass

class SomeDerivedClass(SomeBaseClass):
    """The derived class."""
    async def derived_async_method(self):
        """This is an async method in the derived class."""
        pass
```

and autodoc directives like this:

```
.. automodule:: some_async_module
   :members:
   :inherited-members:
```

The output documentation is like this:

> class some_async_module.SomeBaseClass
> 
>     The base class.
> 
>     await base_async_method()
> 
>         This is an async method in the base class.
> 
> class some_async_module.SomeDerivedClass
> 
>     The derived class.
> 
>     base_async_method()
> 
>         This is an async method in the base class.
> 
>     await derived_async_method()
> 
>         This is an async method in the derived class.

### PR DESCRIPTION
**Pull request by DevilXD**
The cause has been detected and explained by me in #19. The issue was that accessing the `__dict__` on a class directly, didn't resolve MRO entires in any way, so it was just returning `None` for any inherited method (since those are normally accessed after resolving).

Please note that I've ran exactly 0 tests for this, other than trying it out on my documentation (it works beautifully there), but you can deduce from the code that this change shouldn't break anything (that wasn't already broken).

Fixes #19.

### CODE REVIEW

**Comment by DevilXD** [comment]
Does this need anything else to be merged?

**Comment by DevilXD** [comment]
I'd like to note that, after looking into it more, the `inspect` module appears to have a ready-made solution for this problem - it's called `inspect.getattr_static`, and this is roughly how it works: https://i.imgur.com/50x2HXz.png

This should return static and class methods, property descriptors, as well as properly resolve the MRO entries all in one. Could be worth trying.

**Comment by njsmith** [comment]
It would be great if you could also add a test to make sure this doesn't regress in the future. We have an ad hoc testing framework that should make this relatively doable... you would add an inherited async method to `tests/test-docs-source/autodoc_examples.py`, and then in `tests/test-docs-source/test.rst` you'd add something like:

```
.. note::

   .. autoclass:: SubclassWithInheritedAsyncMember
      :members:

   .. code-block:: none

      <em class="property">await </em><code class="(sig-name )?descname">inherited_asyncmethod</code>
```

(The first part invokes autodoc to document `SubclassWithInheritedAsyncMember`, the second part gets converted into a regex that we check against sphinx's generated rendering of the first part, looking for the text `await inherited_asyncmethod`.)

> `inspect.getattr_static`

Oh huh, that's interesting. If that's easier/more reliable and still passes the tests then I'd be fine with switching to it.

**Comment by DevilXD** [comment]
Yeah, so regarding the things that happened here:

• Changed the main portion of the code to use `inspect.getattr_static` instead of the for loop over MRO entries. 
• Rewrote tests to use `inspect.getattr_static` instead of a dict lookup.
• Git-ignored unimportant cache and settings folders/files, related to IDE and liters installed only.
• Standardized the code to be more PEP8 compliant (unused imports, imports sorting, missing empty lines, line length, formatting, `cls` instead of `self` for classmethods, catch explicit exception).
• The mixin class now inherits from `PyObject` - this changes nothing at runtime since the mixing class will come first in the MRO entries regardless, while silencing MyPy complaining about the class having no `self.options` attribute.
• Rewrote a bunch of attribute assignments from `inst.attr = ...` to `setattr(inst, "attr", ...)` form - while both technically work, the second one silences MyPy which doesn't like setting new attributes anywhere outside of the class' init.
• Added tests to the documentation as requested.

To summarize, the code is now relatively cleaner, tests do the same thing as before, and the bug is fixed.

**Comment by DevilXD** [comment]
After some considerations, I've decided to revert the `setattr` usage change. The reason was that it made the code less readable (due to it not being used very often in Python environments), and I guess maintainability in the long run counts here. Each line got a comment silencing MyPy instead.

File: tests/test-docs-source/test.rst [hunk 1]
```diff
@@ -279,6 +279,29 @@ Autodoc + order by source:
...
+      :undoc-members:
+      :inherited-members:
+
+   .. code-block:: none
+
+      <code class="(sig-name )?descname">a_syncmethod</code>
+
+   .. code-block:: none
+
+      <code class="(sig-name )?descname">b_syncmethod</code>
```
Comment (njsmith) [inline tests/test-docs-source/test.rst]: Is this intended to check that they're being rendered as sync functions? Unfortunately I don't think it accomplishes that, because that regex will match regardless of whether there's an `await` earlier on the line.

You could extend the regex to match some extra text before the initial `<code`, or you could use a "negative test", where we assert that the regex does *not* match. Unfortunately you can't do this on a regex-by-regex basis; instead, you do it by using a `.. warning::` block instead of a `.. note::` block. So you'd need to remove these here, and then add something below like:

```
.. warning::

   .. autoclass:: ExampleInheritedSubclass

   .. code-block:: none

      <em class="property">await </em><code class="(sig-name )?descname">a_syncmethod</code>

   .. code-block:: none

      <em class="property">await </em><code class="(sig-name )?descname">b_syncmethod</code>
```

Comment (DevilXD) [reply tests/test-docs-source/test.rst]: Ah, I see. Changed in the next commit.

**Comment by DevilXD** [comment]
The cleanups were done solely because I don't like to work on files where half of the lines are highlighted in red :stuck_out_tongue: The very first commit was done solely through GitHub's edit file functionality, because I didn't want to bother cloning and editing the repo locally, for just a simple change like that. I also figured since most of those were just empty lines and comments added, it should be not that big of a deal. I guess this also answers the second question - I'm not really looking to be a maintainer here, it was just related to making my VSCode not scream at me from half of the lines in the file.

Using MyPy enforces a coding standard of sorts, that semi-guarantees the code won't break, even before you'd try to run it. As you can see from the changes made, it requires doing some things explicitly (adding ignore comments or inheriting from a base class), even tho they'd never really matter at runtime. It's more work, but I think overall it's worth using it.

As a final note here, I'd like to mention that testing the last commit was kinda unsuccessful - I've tried using the below lines to see if the test fails, and it passed in both cases:

```
<code class="(sig-name )?descname">a_syncmethod</code>
<em class="property">await </em><code class="(sig-name )?descname">a_syncmethod</code>
```

Not sure why this might be.

**Comment by DevilXD** [comment]
It might be related due to how the HTML test is constructed. Consider the following case:

```py
for check in checks:
    try:
        assert ...
    except AssertionError:
        ...
        raise
```

If `checks` here would consist of at least 2 different cases, but the first one would raise an `AssertionError`, the whole for loop will exit early, without ever checking the second case. Similarly, if the first check would pass (which is wrong for a negative test), but the second one would fail, the entire test would assume that second `AssertionError` raised was enough to pass the test, which obviously shouldn't be a thing.

It looks like negative tests impose a restriction of one code block test per warning admonition. This was never stated anywhere in the file, and fortunately was never a problem (all negative tests so far, not counting the one I added in the last commit, had only a single code block). I guess this could use at least a comment left somewhere about it.

**Comment by DevilXD** [comment]
Hopefully this should be ok now. Redesigning the test could remove this restriction, but for now, a comment like that should do.

**Comment by njsmith** [comment]
> It looks like negative tests impose a restriction of one code block test per warning admonition. This was never stated anywhere in the file, and fortunately was never a problem (all negative tests so far, not counting the one I added in the last commit, had only a single code block). I guess this could use at least a comment left somewhere about it.

Ohh, nice catch! That's totally an oversight I made when writing that.

It looks like it woudl be pretty easy to fix – instead of expecting `do_html_test` to raise `AssertionError`, we could make it

```python
def do_html_test(node, *, expect_match):
   ...
   if expect_match:
       assert re.search(check, test_content) is not None
   else:
       assert re.search(check, test_content) is None
```

and then call `do_html_test(node, expect_match=False)` / `do_html_test(node, expect_match=True)` down below, as appropriate.

If you're up for making this change, I think it's a good idea – the comment is easy to miss, and it's better not to leave pitfalls like this lying around where someone might fall in later.

**Comment by DevilXD** [comment]
I've reverted the comment commit then, and implemented the test as requested.

**Comment by DevilXD** [comment]
So, well... Does this need anything else to be merged?

**Comment by njsmith** [comment]
Whoops, thanks for the reminder. Looks great!

**Comment by njsmith** [comment]
@DevilXD and this fix is now released to pypi
